runtime/common/src/precompile/honzon.rs

```diff
@@ -0,0 +1,523 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use super::{
+	input::{Input, InputPricer, InputT, Output},
+	target_gas_limit,
+};
+use crate::WeightToGas;
+use frame_support::{log, traits::Get};
+use module_evm::{
+	precompiles::Precompile,
+	runner::state::{PrecompileFailure, PrecompileOutput, PrecompileResult},
+	Context, ExitError, ExitRevert, ExitSucceed,
+};
+use module_honzon::WeightInfo;
+use module_support::HonzonManager;
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+use primitives::{Amount, Balance, CurrencyId, Position};
+use sp_runtime::{traits::Convert, FixedPointNumber, RuntimeDebug};
+use sp_std::{marker::PhantomData, prelude::*};
+
+/// The Honzon precomnpile
+///
+/// `input` data starts with `action`.
+///
+/// Actions:
+///  - Adjust loan. `input` bytes: `who`, `currency_id`, `collateral_adjustment`,
+///    `debit_adjustment`.
+///  - Close loan by dex. `input` bytes: `who`, `currency_id`, `max_collateral_amount`.
+///  - Get position. `input` bytes: `who`, `currency_id`.
+///  - Get liquidation ratio. `input` bytes: `currency_id`.
+///  - Get current collateral ratio. `input` bytes: `who`, `currency_id`.
+pub struct HonzonPrecompile<R>(PhantomData<R>);
+
+#[module_evm_utility_macro::generate_function_selector]
+#[derive(RuntimeDebug, Eq, PartialEq, TryFromPrimitive, IntoPrimitive)]
+#[repr(u32)]
+pub enum Action {
+	AdjustLoan = "adjustLoan(address,address,int256,int256)",
+	CloseLoanByDex = "closeLoanByDex(address,address,uint256)",
+	GetPosition = "getPosition(address,address)",
+	GetLiquidationRatio = "getLiquidationRatio(address)",
+	GetCurrentCollateralRatio = "getCurrentCollateralRatio(address,address)",
+}
+
+impl<Runtime> Precompile for HonzonPrecompile<Runtime>
+where
+	Runtime: module_evm::Config + module_honzon::Config + module_prices::Config,
+	module_honzon::Pallet<Runtime>: HonzonManager<Runtime::AccountId, CurrencyId, Amount, Balance>,
+{
+	fn execute(input: &[u8], target_gas: Option<u64>, _context: &Context, _is_static: bool) -> PrecompileResult {
+		let input = Input::<Action, Runtime::AccountId, Runtime::AddressMapping, Runtime::Erc20InfoMapping>::new(
+			input,
+			target_gas_limit(target_gas),
+		);
+
+		let gas_cost = Pricer::<Runtime>::cost(&input)?;
+
+		if let Some(gas_limit) = target_gas {
+			if gas_limit < gas_cost {
+				return Err(PrecompileFailure::Error {
+					exit_status: ExitError::OutOfGas,
+				});
+			}
+		}
+
+		let action = input.action()?;
+
+		match action {
+			Action::AdjustLoan => {
+				let who = input.account_id_at(1)?;
+				let currency_id = input.currency_id_at(2)?;
+				let collateral_adjustment = input.i128_at(3)?;
+				let debit_adjustment = input.i128_at(4)?;
```

Comment:
`adjustLoan(address,address,int256,int256)`
It's `int256`

Reply:
@zjb0807 I suggest changing action to `adjustLoan(address,address,int128,int128)
`